### ISSUE TITLE
Mutate thrown error "subs[i] not a function."

### ISSUE THREAD

**Issue by Ogzhnsfgl** [vercel/swr#2357]
# Bug report

## Description / Observed Behavior

Hi everyone,

I upgraded SWR from 1.2.2 to 2.0.0. After upgrading, some cases are broken. Mutate function thrown error 'subs[i] not a function.' If I refresh the app completely, there is no error thrown. 

Example case:

1) I have BasketData fetched SWR.
2) I send a request to another service to remove a basket item and then mutate the basketData loading state manually.
3) When the request is done I mutate basketData with the response. In this step, the error is thrown.

If I do the same flow after f5 (reload app), there is no error.

## Expected Behavior

I rollback to SWR 1.2.2 and there is no issue there. I expect the same behavior with 2.0.0.

## Repro Steps / Code Example

Or share your code snippet or a [CodeSandbox](https://codesandbox.io) link is also appreciated!

## Additional Context

SWR version 2.0.0

![Screen Shot 2023-01-11 at 09 31 53](https://user-images.githubusercontent.com/85337758/211734405-8f2e91bf-d4cf-4ee2-a52e-93fd3c2898bb.png)

**Comment by lfbergee** [comment]
We also experience this intermittently. It's quite hard to find a stable way to reproduce, however we see in our Sentry logs it does happen a fair bit across all browsers and devices. I estimate it happens on every 30 refresh, give or take. On that page is just one hook using SWR. That hook is used by a couple of components.

```
export const usePages = () => {
  const { data, error } = useSWRImmutable(url, async url => {
    const res = await request({ url });
    return res ? mapResponseToShallowPages(res) : undefined;
  });

  return { pages: data, defaultPage: data?.[0] ?? undefined, error };
};
```

It looks like it doesn't crash the application, and as far as I can see it does successfully fetch what its supposed to. Don't really know where to keep digging here.

We've added this patch for now, but it would be super interesting to find out why this occurs.

```
+ 45
- 0
Viewed
diff --git a/node_modules/swr/_internal/dist/index.esm.js b/node_modules/swr/_internal/dist/index.esm.js
index 524b14f..b98ed73 100644
--- a/node_modules/swr/_internal/dist/index.esm.js
+++ b/node_modules/swr/_internal/dist/index.esm.js
@@ -398,7 +398,9 @@ const initCache = (provider, options)=>{
             const subs = subscriptions[key];
             if (subs) {
                 for(let i = subs.length; i--;){
-                    subs[i](value, prev);
+                    if(typeof subs[i] === 'function'){
+                        subs[i](value, prev);
+                    }
                 }
             }
         };
```

If this patch looks like it make sense to you, I'll be happy to create a PR.

**Comment by MadratJerry** [comment]
> We also experience this intermittently. It's quite hard to find a stable way to reproduce, however we see in our Sentry logs it does happen a fair bit across all browsers and devices. I estimate it happens on every 30 refresh, give or take. On that page is just one hook using SWR. That hook is used by a couple of components.
> 
> ```
> export const usePages = () => {
>   const { data, error } = useSWRImmutable(url, async url => {
>     const res = await request({ url });
>     return res ? mapResponseToShallowPages(res) : undefined;
>   });
> 
>   return { pages: data, defaultPage: data?.[0] ?? undefined, error };
> };
> ```
> 
> It looks like it doesn't crash the application, and as far as I can see it does successfully fetch what its supposed to. Don't really know where to keep digging here.
> 
> We've added this patch for now, but it would be super interesting to find out why this occurs.
> 
> ```
> + 45
> - 0
> Viewed
> diff --git a/node_modules/swr/_internal/dist/index.esm.js b/node_modules/swr/_internal/dist/index.esm.js
> index 524b14f..b98ed73 100644
> --- a/node_modules/swr/_internal/dist/index.esm.js
> +++ b/node_modules/swr/_internal/dist/index.esm.js
> @@ -398,7 +398,9 @@ const initCache = (provider, options)=>{
>              const subs = subscriptions[key];
>              if (subs) {
>                  for(let i = subs.length; i--;){
> -                    subs[i](value, prev);
> +                    if(typeof subs[i] === 'function'){
> +                        subs[i](value, prev);
> +                    }
>                  }
>              }
>          };
> ```
> 
> If this patch looks like it make sense to you, I'll be happy to create a PR.

I ran into the same problem, and like you, it was very difficult to reproduce. But I've done it locally again with some quick toggle clicks to trigger the request

**Comment by lfbergee** [comment]
I couldn't let this go, so I've kept digging.

It looks to me as the `subscribe` function can be called and mutate the subscription-object after the `initCache` function sets up the loop, but before it finishes.

So in my case the `subscriptions[key]` object has either 3 or 5 entries, somewhat randomly. That array contains anonymous functions, so not that easy to decipher why it flips between 3 and 5 entries. And in particular cases it starts out with 5, then gets mutated to have 3 entries. But the for-loop has already been initiated with length 5. That results in index out of bounds. 

If `const subs ` made a copy instead of referencing `subscriptions[key]`, or the iteration were done on the array instead of a for-loop, this could be avoided.

**Comment by 7michele7** [comment]
We just updated to `2.0.3` from `1.x` and we are facing the same issue. We couldn't find a way to reproduce it yet.

<img width="912" alt="Screenshot 2023-02-07 at 22 32 30" src="https://user-images.githubusercontent.com/17926167/217382190-68352486-48eb-4757-a92c-726f5f9ee37f.png">

**Comment by promer94** [comment]
@7michele7 @lfbergee @Ogzhnsfgl @MadratJerry 

Sorry to borther you folks. 
But could you share me your version of react here ? I am try to reproduce this problem and create a test case for this

**Comment by Ogzhnsfgl** [comment]
> @7michele7 @lfbergee @Ogzhnsfgl @MadratJerry
> 
> Sorry to borther you folks. But could you share me your version of react here ? I am try to reproduce this problem and create a test case for this

@promer94 

`    "react": "17.0.2",
    "react-dom": "17.0.2",
`

**Comment by Ogzhnsfgl** [comment]
In my case, I solved the problem by changing my useSWR style. I'll try to explain:

- In my project, there are no global state management packages, if required we use the cache of SWR. So, we have a lot of calls for useSWR.

For example, we need basket data on the basket page and we call this function shown below multiple times on the same page

`
const {data,mutate,isValidating} = useSWR('api/basket')
`


Attention here: We were calling that function from different component on **the same page**. Some of these components render conditionally.

We try to avoid passing the prop with the cache feature of SWR but it causes that bug. So that we change our code different approach. We call useSWR at the top of the page and pass all necessary functions and data to the children with props or context API.

**Comment by 7michele7** [comment]
Sure @promer94 , thanks for digging into this.

```
"react": "^16.13.0"
"react-router-dom": "^5.1.2",
```

**Comment by promer94** [comment]
We will keep this open until we find a stable way to reproduce and have a test for this

**Comment by huozhi** [comment]
@Ogzhnsfgl  `swr@2.0.4` is released with the fix, can you confirm if that fix your issue?

**Comment by SandroMaglione** [comment]
@huozhi upgrading from `v2.0.0` to `v2.0.4` fixed the error for me

**Comment by 7michele7** [comment]
The issue is resolved for us too, with `v2.0.4`. Thanks y'all.

**Comment by huozhi** [comment]
Thanks for the confirmation! 🙏
